File: src/domain/fonctionnalites/autofocus_service.py
import asyncio
import re
import threading
from collections.abc import Callable

from src.domain.constants import AUTOFOCUS_POLL_INTERVAL
from src.domain.fonctionnalites.title_parser import extract_pseudo_from_title
from src.domain.fonctionnalites.window_service import WindowService
from src.domain.utils.win32_helpers import WIN32_OK, WINSDK_OK, copy_text_to_clipboard, winman, winnot

NOTIFICATION_PATTERNS = [
    ("combat", ("de jouer",), "combat"),
    ("echange", ("te propose de faire un echange",), "echange"),
    ("groupe", ("t'invite", "rejoindre son groupe"), "groupe"),
    ("mp", ("de ", "(prive)"), "mp"),
]

MP_SENDER_PATTERN = re.compile(
    r"^\s*(?:de\s+|\((?:prive|priv\u00e9)\)\s+)(?P<sender>[^:]+?)\s*:",
    re.IGNORECASE,
)
UNITY_NOTIFICATION_TITLE_PATTERN = re.compile(r"^dofus\s+3(?:\.\d+)*$", re.IGNORECASE)


def _normalize_text(value: str) -> str:
    return (
        value.replace("\u00e9", "e")
        .replace("\u00e8", "e")
        .replace("\u00ea", "e")
        .replace("\u00e0", "a")
        .replace("\u00e2", "a")
        .replace("\u00f9", "u")
        .replace("\u00fb", "u")
        .replace("\u00ef", "i")
        .replace("\u00ee", "i")
        .replace("\u00f4", "o")
        .replace("\u00e7", "c")
        .casefold()
    )


class AutoFocusService:
    def __init__(
        self,
        window_service: WindowService,
        log_fn: Callable[[str, str], None],
        stats_fn: Callable[[dict[str, str]], None],
    ) -> None:
        self._window_service = window_service
        self._log_fn = log_fn
        self._stats_fn = stats_fn
        self._thread: threading.Thread | None = None
        self._loop: asyncio.AbstractEventLoop | None = None
        self._running = False
        self._debug_enabled = False
        self._enabled_types = {
            "combat": True,
            "echange": True,
            "groupe": True,
            "mp": True,
        }
        self._stats = {
            "notifications": "0",
            "matches": "0",
            "focus": "0",
            "last": "-",
        }
        self._filters_getter: Callable[[], tuple[bool, bool]] = lambda: (True, True)
        self._character_rule_getter: Callable[[str, str], bool] = lambda _pseudo, _notif_type: True
        self._mp_clipboard_enabled_getter: Callable[[], bool] = lambda: True

    @property
    def available(self) -> bool:
        return WIN32_OK and WINSDK_OK

    @property
    def running(self) -> bool:
        return self._running

    def set_filters_getter(self, getter: Callable[[], tuple[bool, bool]]) -> None:
        self._filters_getter = getter

    def set_character_rule_getter(self, getter: Callable[[str, str], bool]) -> None:
        self._character_rule_getter = getter

    def set_mp_clipboard_enabled_getter(self, getter: Callable[[], bool]) -> None:
        self._mp_clipboard_enabled_getter = getter

    def set_debug(self, enabled: bool) -> None:
        self._debug_enabled = enabled

    def set_type_enabled(self, key: str, enabled: bool) -> None:
        if key in self._enabled_types:
            self._enabled_types[key] = enabled

    def is_type_enabled(self, key: str) -> bool:
        return self._enabled_types.get(key, False)

    def start(self) -> None:
        if not self.available or self._running:
            return
        self._running = True
        self._thread = threading.Thread(target=self._run_async_loop, daemon=True)
        self._thread.start()

    def stop(self) -> None:
        self._running = False
        if self._loop is not None and self._loop.is_running():
            self._loop.call_soon_threadsafe(self._loop.stop)

    def _run_async_loop(self) -> None:
        self._loop = asyncio.new_event_loop()
        asyncio.set_event_loop(self._loop)
        try:
            self._loop.run_until_complete(self._listen())
        except Exception as exc:
            if self._running:
                self._log_fn(f"Erreur fatale AutoFocus : {exc}", "error")
        finally:
            self._loop.close()

    async def _listen(self) -> None:
        listener = winman.UserNotificationListener.current
        access = await listener.request_access_async()
        if access != winman.UserNotificationListenerAccessStatus.ALLOWED:
            self._log_fn(
                "Acces notifications refuse. Active-les dans Parametres -> Systeme -> Notifications.",
                "error",
            )
            self.stop()
            return

        self._log_fn("Acces aux notifications accorde.", "ok")
        seen_ids: set[int] = set()
        change_event = asyncio.Event()
        token = None
        use_events = False

        def on_notification_changed(_sender, _args) -> None:
            if self._loop is not None and self._loop.is_running():
                self._loop.call_soon_threadsafe(change_event.set)

        try:
            token = listener.add_notification_changed(on_notification_changed)
            use_events = True
            self._log_fn("Mode event-driven actif.", "ok")
        except Exception:
            self._log_fn("Mode polling actif.", "dim")

        try:
            while self._running:
                if use_events:
                    try:
                        await asyncio.wait_for(change_event.wait(), timeout=30.0)
                    except asyncio.TimeoutError:
                        pass
                    except asyncio.CancelledError:
                        break
                    change_event.clear()
                else:
                    try:
                        await asyncio.sleep(AUTOFOCUS_POLL_INTERVAL)
                    except asyncio.CancelledError:
                        break

                notifications = await listener.get_notifications_async(
                    winnot.NotificationKinds.TOAST,
                )
                new_notifications = [item for item in notifications if item.id not in seen_ids]
                if not new_notifications:
                    continue

                self._increment_stat("notifications", len(new_notifications))

                for notification in new_notifications:
                    seen_ids.add(notification.id)
                    await self._handle_notification(notification)

                if len(seen_ids) > 500:
                    seen_ids.clear()
        finally:
            try:
                listener.remove_notification_changed(token)
            except Exception:
                pass

    async def _handle_notification(self, notification) -> None:
        try:
            binding = notification.notification.visual.get_binding(
                winnot.KnownNotificationBindings.toast_generic,
            )
        except Exception:
            binding = None

        if binding is None:
            return

        elements = [element.text for element in binding.get_text_elements()]
        if not elements:
            return

        title = elements[0]
        body = " | ".join(part.strip() for part in elements[1:] if part and part.strip())

        if self._debug_enabled:
            self._log_fn(f"[debug] titre={title!r} corps={body!r}", "debug")

        notif_type = self._detect_notification_type(body)
        if notif_type is None:
            return

        if notif_type == "mp" and self._mp_clipboard_enabled_getter():
            sender = self._extract_mp_sender(body)
            if sender:
                whisper = f"/w {sender} "
                if copy_text_to_clipboard(whisper):
                    self._log_fn(f"  Presse-papiers -> {whisper}", "ok")
                else:
                    self._log_fn("  Echec copie presse-papiers MP.", "warn")

        pseudo = extract_pseudo_from_title(title)
        unity_fallback = pseudo is None and self._is_unity_notification_title(title)
        if unity_fallback and self._window_service.count_game_windows("unity") > 1:
            self._log_fn("[unity] ignore focus (plusieurs fenetres Unity ouvertes)", "dim")
            return
        if pseudo is None and not unity_fallback:
            return

        target_label = pseudo or "Unity"

        if not self.is_type_enabled(notif_type):
            self._log_fn(f"[{notif_type}] ignore (desactive) -> {target_label}", "dim")
            return

        if pseudo and not self._character_rule_getter(pseudo, notif_type):
            self._log_fn(f"[{notif_type}] ignore pour {pseudo} (regle personnage)", "dim")
            return

        self._increment_stat("matches", 1)
        self._stats["last"] = target_label
        self._stats_fn(dict(self._stats))
        self._log_fn(f"[{notif_type.upper()}] {target_label} -> {body}", f"type_{notif_type}")

        enable_retro, enable_unity = self._filters_getter()
        if pseudo:
            ok, detail = self._window_service.focus_pseudo(
                pseudo,
                enable_retro=enable_retro,
                enable_unity=enable_unity,
            )
        else:
            ok, detail = self._window_service.focus_game_type(
                "unity",
                enable_retro=enable_retro,
                enable_unity=enable_unity,
            )

        if ok:
            self._increment_stat("focus", 1)
            self._log_fn(f"  Focus -> {detail}", "ok")
        else:
            self._log_fn(f"  {detail}", "error")

    def _increment_stat(self, key: str, increment: int) -> None:
        current = int(self._stats[key])
        self._stats[key] = str(current + increment)
        self._stats_fn(dict(self._stats))

    @staticmethod
    def _detect_notification_type(body: str) -> str | None:
        normalized_body = _normalize_text(body)
        for key, patterns, _label in NOTIFICATION_PATTERNS:
            if any(pattern in normalized_body for pattern in patterns):
                return key
        return None

    @staticmethod
    def _is_unity_notification_title(title: str) -> bool:
        return UNITY_NOTIFICATION_TITLE_PATTERN.fullmatch(title.strip()) is not None

    @staticmethod
    def _extract_mp_sender(body: str) -> str | None:
        match = MP_SENDER_PATTERN.match(body.strip())
        if match is None:
            return None
        sender = match.group("sender").strip()
        return sender or None

File: dofus-helper/backend/src/services/autofocusService.ts
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname } from "node:path";
import { ensureCharacterRule, getCharactersConfig, type CharacterNotifType } from "./charactersStorage.js";
import { countGameWindows, extractPseudoFromTitle, focusGameType, focusPseudo, isUnityNotificationTitle, listGameWindows } from "./gameWindows.js";
import { fetchToastNotifications } from "./notificationsBridge.js";
import { resolveDataFile } from "./runtimePaths.js";
import { copyTextToClipboard } from "./windowsBridge.js";

export interface AutofocusLogEntry {
  id: number;
  timestamp: string;
  message: string;
  tag: string;
}

export interface AutofocusStats {
  notifications: string;
  matches: string;
  focus: string;
  last: string;
}

export interface AutofocusState {
  available: boolean;
  running: boolean;
  debugEnabled: boolean;
  enabledTypes: Record<CharacterNotifType, boolean>;
  stats: AutofocusStats;
  accessStatus: string;
  logs?: AutofocusLogEntry[];
}

interface PersistedAutofocusConfig {
  debugEnabled: boolean;
  enabledTypes: Record<CharacterNotifType, boolean>;
}

const POLL_INTERVAL_MS = 900;
const MAX_LOGS = 300;
const MAX_SEEN_IDS = 500;
const NOTIFICATION_PATTERNS: Array<{ type: CharacterNotifType; patterns: string[] }> = [
  { type: "combat", patterns: ["de jouer"] },
  { type: "echange", patterns: ["te propose de faire un echange"] },
  { type: "groupe", patterns: ["t'invite", "t'invite a rejoindre son groupe", "rejoindre son groupe", "invite a rejoindre son groupe"] },
  { type: "mp", patterns: ["de ", "(prive)"] },
];
const MP_SENDER_PATTERN = /^\s*(?:de\s+|\((?:prive|privé)\)\s+)(?<sender>[^:]+?)\s*:/i;

function defaultEnabledTypes(): Record<CharacterNotifType, boolean> {
  return {
    combat: true,
    echange: true,
    groupe: true,
    mp: true,
  };
}

function defaultPersistedConfig(): PersistedAutofocusConfig {
  return {
    debugEnabled: false,
    enabledTypes: defaultEnabledTypes(),
  };
}

function defaultStats(): AutofocusStats {
  return {
    notifications: "0",
    matches: "0",
    focus: "0",
    last: "-",
  };
}

function resolveStorageFilePath(): string {
  return resolveDataFile("autofocus.json");
}

function normalizePersistedConfig(value: unknown): PersistedAutofocusConfig {
  if (typeof value !== "object" || value === null) {
    return defaultPersistedConfig();
  }
  const candidate = value as Record<string, unknown>;
  const enabledTypes = defaultEnabledTypes();
  for (const type of Object.keys(enabledTypes) as CharacterNotifType[]) {
    enabledTypes[type] = candidate.enabledTypes && typeof candidate.enabledTypes === "object"
      ? ((candidate.enabledTypes as Record<string, unknown>)[type] !== false)
      : true;
  }
  return {
    debugEnabled: candidate.debugEnabled === true,
    enabledTypes,
  };
}

function readPersistedConfig(): PersistedAutofocusConfig {
  const filePath = resolveStorageFilePath();
  if (!existsSync(filePath)) {
    return defaultPersistedConfig();
  }
  try {
    return normalizePersistedConfig(JSON.parse(readFileSync(filePath, "utf8")));
  } catch {
    return defaultPersistedConfig();
  }
}

function writePersistedConfig(config: PersistedAutofocusConfig): PersistedAutofocusConfig {
  const filePath = resolveStorageFilePath();
  mkdirSync(dirname(filePath), { recursive: true });
  writeFileSync(filePath, JSON.stringify(config, null, 2), "utf8");
  return normalizePersistedConfig(config);
}

function timestamp(): string {
  return new Date().toLocaleTimeString("fr-FR", { hour12: false });
}

function normalizeText(value: string): string {
  return value
    .replace(/[’`´]/g, "'")
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "")
    .toLowerCase();
}

function detectNotificationType(body: string): CharacterNotifType | null {
  const normalizedBody = normalizeText(body);
  for (const entry of NOTIFICATION_PATTERNS) {
    if (entry.patterns.some((pattern) => normalizedBody.includes(normalizeText(pattern)))) {
      return entry.type;
    }
  }
  return null;
}

function extractMpSender(body: string): string | null {
  const match = MP_SENDER_PATTERN.exec(body.trim());
  const sender = match?.groups?.sender?.trim() ?? "";
  return sender || null;
}

class AutoFocusService {
  private persistedConfig = readPersistedConfig();
  private logs: AutofocusLogEntry[] = [];
  private logSeq = 0;
  private stats: AutofocusStats = defaultStats();
  private seenNotificationIds = new Set<number>();
  private intervalHandle: ReturnType<typeof setInterval> | null = null;
  private pollInFlight = false;
  private available = true;
  private running = false;
  private accessStatus = "Unknown";
  private lastErrorMessage = "";
  private lastAccessWarning = "";

  constructor() {
    this.syncRunningState(false);
  }

  getState(includeLogs: boolean): AutofocusState {
    return {
      available: this.available,
      running: this.running,
      debugEnabled: this.persistedConfig.debugEnabled,
      enabledTypes: { ...this.persistedConfig.enabledTypes },
      stats: { ...this.stats },
      accessStatus: this.accessStatus,
      logs: includeLogs ? [...this.logs] : undefined,
    };
  }

  poll(lastLogId: number): { logs: AutofocusLogEntry[]; state: AutofocusState } {
    return {
      logs: this.logs.filter((entry) => entry.id > lastLogId),
      state: this.getState(false),
    };
  }

  setTypeEnabled(notifType: string, enabled: boolean): AutofocusState {
    if (!Object.prototype.hasOwnProperty.call(this.persistedConfig.enabledTypes, notifType)) {
      throw new Error("Type AutoFocus invalide.");
    }
    this.persistedConfig.enabledTypes[notifType as CharacterNotifType] = enabled;
    this.persistedConfig = writePersistedConfig(this.persistedConfig);
    this.appendLog(`[${notifType}] ${enabled ? "active" : "desactive"}.`, enabled ? `type_${notifType}` : "dim");
    this.syncRunningState(true);
    return this.getState(false);
  }

  setDebug(enabled: boolean): AutofocusState {
    this.persistedConfig.debugEnabled = enabled;
    this.persistedConfig = writePersistedConfig(this.persistedConfig);
    this.appendLog(enabled ? "Mode debug AutoFocus actif." : "Mode debug AutoFocus desactive.", enabled ? "debug" : "dim");
    this.syncRunningState(false);
    return this.getState(false);
  }

  private hasEnabledType(): boolean {
    return Object.values(this.persistedConfig.enabledTypes).some(Boolean);
  }

  private syncRunningState(logOnStartStop: boolean): void {
    if (this.hasEnabledType()) {
      if (!this.running) {
        this.start(logOnStartStop);
      }
      return;
    }
    if (this.running) {
      this.stop(logOnStartStop ? "AutoFocus stoppee (aucun type actif)." : null);
    }
  }

  private start(logMessage: boolean): void {
    if (this.running) {
      return;
    }
    this.running = true;
    if (logMessage) {
      this.appendLog("AutoFocus active.", "ok");
    }
    void this.pollNotifications();
    this.intervalHandle = setInterval(() => {
      void this.pollNotifications();
    }, POLL_INTERVAL_MS);
  }

  private stop(message: string | null): void {
    if (this.intervalHandle) {
      clearInterval(this.intervalHandle);
      this.intervalHandle = null;
    }
    this.running = false;
    if (message) {
      this.appendLog(message, "dim");
    }
  }

  private appendLog(message: string, tag: string): void {
    this.logSeq += 1;
    this.logs.push({
      id: this.logSeq,
      timestamp: timestamp(),
      message,
      tag,
    });
    if (this.logs.length > MAX_LOGS) {
      this.logs.splice(0, this.logs.length - MAX_LOGS);
    }
  }

  private incrementStat(key: keyof AutofocusStats, increment: number): void {
    const next = Number.parseInt(this.stats[key], 10) + increment;
    this.stats[key] = String(Number.isFinite(next) ? next : increment);
  }

  private async pollNotifications(): Promise<void> {
    if (!this.running || this.pollInFlight) {
      return;
    }

    this.pollInFlight = true;
    try {
      const snapshot = await fetchToastNotifications();
      this.available = true;
      this.accessStatus = snapshot.access;

      if (snapshot.access !== "Allowed") {
        const nextMessage = snapshot.access === "Denied"
          ? "Acces notifications refuse. Active-les dans Parametres -> Systeme -> Notifications."
          : snapshot.access === "Unknown"
            ? "Acces notifications en attente."
            : `Acces notifications indisponible: ${snapshot.access}.`;
        if (nextMessage !== this.lastAccessWarning) {
          this.lastAccessWarning = nextMessage;
          this.appendLog(nextMessage, snapshot.access === "Denied" ? "error" : "warn");
        }
        return;
      }

      this.lastAccessWarning = "";
      this.lastErrorMessage = "";

      if (this.persistedConfig.debugEnabled && snapshot.notifications.length > 0) {
        this.appendLog(
          `[debug] snapshot count=${snapshot.notifications.length} ids=${snapshot.notifications.map((item) => item.id).join(",")}`,
          "debug",
        );
      }

      const freshNotifications = snapshot.notifications.filter((item) => !this.seenNotificationIds.has(item.id));
      if (this.persistedConfig.debugEnabled && snapshot.notifications.length > 0) {
        this.appendLog(
          `[debug] fresh count=${freshNotifications.length} ids=${freshNotifications.map((item) => item.id).join(",")}`,
          "debug",
        );
      }
      if (!freshNotifications.length) {
        return;
      }

      this.incrementStat("notifications", freshNotifications.length);
      for (const notification of freshNotifications) {
        this.seenNotificationIds.add(notification.id);
        await this.handleNotification(notification.title, notification.body);
      }
      if (this.seenNotificationIds.size > MAX_SEEN_IDS) {
        this.seenNotificationIds.clear();
      }
    } catch (error) {
      this.available = false;
      this.accessStatus = "Unavailable";
      const message = error instanceof Error ? error.message : "Erreur AutoFocus inconnue.";
      if (message !== this.lastErrorMessage) {
        this.lastErrorMessage = message;
        this.appendLog(`Erreur AutoFocus : ${message}`, "error");
      }
      this.stop(null);
    } finally {
      this.pollInFlight = false;
    }
  }

  private async handleNotification(title: string, body: string): Promise<void> {
    if (this.persistedConfig.debugEnabled) {
      this.appendLog(`[debug] titre=${JSON.stringify(title)} corps=${JSON.stringify(body)}`, "debug");
    }

    const notifType = detectNotificationType(body);
    if (!notifType) {
      return;
    }

    const charactersConfig = getCharactersConfig();

    if (notifType === "mp" && charactersConfig.copyMpSender) {
      const sender = extractMpSender(body);
      if (sender) {
        const whisper = `/w ${sender} `;
        if (await copyTextToClipboard(whisper)) {
          this.appendLog(`Presse-papiers -> ${whisper}`, "ok");
        } else {
          this.appendLog("Echec copie presse-papiers MP.", "warn");
        }
      }
    }

    let pseudo = extractPseudoFromTitle(title);
    const unityFallback = pseudo === null && isUnityNotificationTitle(title);
    if (unityFallback && (await countGameWindows("unity")) > 1) {
      this.appendLog("[unity] ignore focus (plusieurs fenetres Unity ouvertes)", "dim");
      return;
    }
    if (pseudo === null && !unityFallback) {
      return;
    }

    const filters = {
      enableRetro: charactersConfig.enableRetro,
      enableUnity: charactersConfig.enableUnity,
    };

    if (pseudo === null) {
      const unityWindows = (await listGameWindows(filters)).filter((window) => window.gameType === "unity");
      if (unityWindows.length === 1) {
        pseudo = unityWindows[0].pseudo;
      }
    }

    const targetLabel = pseudo ?? "Unity";

    if (!this.persistedConfig.enabledTypes[notifType]) {
      this.appendLog(`[${notifType}] ignore (desactive) -> ${targetLabel}`, "dim");
      return;
    }

    if (pseudo) {
      const rule = ensureCharacterRule(charactersConfig, pseudo);
      if (!(rule.rotation !== false && rule[notifType] !== false)) {
        this.appendLog(`[${notifType}] ignore pour ${pseudo} (regle personnage)`, "dim");
        return;
      }
    }

    this.incrementStat("matches", 1);
    this.stats.last = targetLabel;
    this.appendLog(`[${notifType.toUpperCase()}] ${targetLabel} -> ${body}`, `type_${notifType}`);

    const result = pseudo
      ? await focusPseudo(pseudo, filters)
      : await focusGameType("unity", filters);

    if (result.ok) {
      this.incrementStat("focus", 1);
      this.appendLog(`Focus -> ${result.message}`, "ok");
    } else {
      this.appendLog(result.message, "error");
    }
  }
}

export const autofocusService = new AutoFocusService();

